src/main/java/uk/gov/pay/connector/events/eventdetails/charge/PaymentDetailsTakenFromPaymentInstrumentEventDetails.java

```diff
@@ -0,0 +1,140 @@
+    private final String lastDigitsCardNumber;
+    private final String cardholderName;
+    private final String expiryDate;
+    private final String addressLine1;
+    private final String addressLine2;
+    private final String addressPostcode;
+    private final String addressCity;
+    private final String addressCounty;
+    private final String addressStateProvince;
+    private final String addressCountry;
+
+    PaymentDetailsTakenFromPaymentInstrumentEventDetails(String cardType, String cardBrand, String cardBrandLabel, String firstDigitsCardNumber, String lastDigitsCardNumber, String cardholderName, String expiryDate, String addressLine1, String addressLine2, String addressPostcode, String addressCity, String addressCounty, String addressStateProvince, String addressCountry) {
+        this.cardType = cardType;
+        this.cardBrand = cardBrand;
+        this.cardBrandLabel = cardBrandLabel;
+        this.firstDigitsCardNumber = firstDigitsCardNumber;
+        this.lastDigitsCardNumber = lastDigitsCardNumber;
+        this.cardholderName = cardholderName;
+        this.expiryDate = expiryDate;
+        this.addressLine1 = addressLine1;
+        this.addressLine2 = addressLine2;
+        this.addressPostcode = addressPostcode;
+        this.addressCity = addressCity;
+        this.addressCounty = addressCounty;
+        this.addressStateProvince = addressStateProvince;
+        this.addressCountry = addressCountry;
+    }
+
+    public static PaymentDetailsTakenFromPaymentInstrumentEventDetails from(ChargeEntity charge) {
+        var cardDetails = charge.getCardDetails();
+        var builder = new PaymentDetailsTakenFromPaymentInstrumentEventDetailsBuilder()
+                .setCardType(Optional.ofNullable(cardDetails.getCardType()).map(Enum::toString).orElse(null))
+                .setCardBrand(cardDetails.getCardBrand())
+                .setCardBrandLabel(cardDetails.getCardTypeDetails().map(CardBrandLabelEntity::getLabel).orElse(null))
+                .setFirstDigitsCardNumber(cardDetails.getFirstDigitsCardNumber().toString())
+                .setLastDigitsCardNumber(cardDetails.getLastDigitsCardNumber().toString())
+                .setCardholderName(cardDetails.getCardHolderName())
+                .setExpiryDate(cardDetails.getExpiryDate().toString());
+
+        cardDetails.getBillingAddress().ifPresent(billingAddress -> {
+            builder.setAddressLine1(billingAddress.getLine1())
+                    .setAddressLine2(billingAddress.getLine2())
+                    .setAddressPostcode(billingAddress.getPostcode())
+                    .setAddressCity(billingAddress.getCity())
+                    .setAddressCounty(billingAddress.getCounty())
+                    .setAddressStateProvince(billingAddress.getStateOrProvince())
+                    .setAddressCountry(billingAddress.getCountry());
+        });
+        return builder.createPaymentDetailsTakenFromPaymentInstrumentEventDetails();
```

Comment:
Could pass builder to constructor similar to https://github.com/alphagov/pay-connector/blob/d97c2e860a09385943ea5467c7bf68bb37a4d210/src/main/java/uk/gov/pay/connector/events/eventdetails/charge/PaymentDetailsEnteredEventDetails.java#L37

```suggestion
        return new PaymentDetailsTakenFromPaymentInstrumentEventDetails(builder)
```

Reply:
@kbottla 👍 https://github.com/alphagov/pay-connector/commit/80c080175aba25fbbc8680b976d6d82077c4ea6b